smithy-rules-engine/src/main/java/software/amazon/smithy/rulesengine/traits/OperationContextParamsTraitValidator.java

```diff
@@ -0,0 +1,455 @@
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.OperationIndex;
+import software.amazon.smithy.model.shapes.BigDecimalShape;
+import software.amazon.smithy.model.shapes.BigIntegerShape;
+import software.amazon.smithy.model.shapes.BlobShape;
+import software.amazon.smithy.model.shapes.BooleanShape;
+import software.amazon.smithy.model.shapes.ByteShape;
+import software.amazon.smithy.model.shapes.DocumentShape;
+import software.amazon.smithy.model.shapes.DoubleShape;
+import software.amazon.smithy.model.shapes.FloatShape;
+import software.amazon.smithy.model.shapes.IntegerShape;
+import software.amazon.smithy.model.shapes.ListShape;
+import software.amazon.smithy.model.shapes.LongShape;
+import software.amazon.smithy.model.shapes.MapShape;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ResourceShape;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeVisitor;
+import software.amazon.smithy.model.shapes.ShortShape;
+import software.amazon.smithy.model.shapes.StringShape;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.shapes.TimestampShape;
+import software.amazon.smithy.model.shapes.UnionShape;
+import software.amazon.smithy.model.traits.LengthTrait;
+import software.amazon.smithy.model.traits.RangeTrait;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.ValidationEvent;
+import software.amazon.smithy.utils.ListUtils;
+import software.amazon.smithy.utils.SmithyUnstableApi;
+
+/**
+ * Validates {@link OperationContextParamsTrait} traits.
+ */
+@SmithyUnstableApi
+public final class OperationContextParamsTraitValidator extends AbstractValidator {
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        ContextIndex index = ContextIndex.of(model);
+        List<ValidationEvent> events = new ArrayList<>();
+        for (OperationShape operationShape : model.getOperationShapes()) {
+            Map<String, OperationContextParamDefinition> definitionMap = index.getOperationContextParams(operationShape)
+                    .map(OperationContextParamsTrait::getParameters)
+                    .orElse(Collections.emptyMap());
+            for (Map.Entry<String, OperationContextParamDefinition> entry : definitionMap.entrySet()) {
+
+                try {
+                    JmespathExpression path = JmespathExpression.parse(entry.getValue().getPath());
+                    StructureShape input = OperationIndex.of(model).expectInputShape(operationShape);
+                    LinterResult linterResult = path.lint(createCurrentNodeFromShape(input, model));
+
+                    if (!linterResult.getProblems().isEmpty()) {
```

Comment:
Can we add some tests for the linter result?

E.g.,

* Using an array as object, `someArray.key`
* Using an object as array, `someObject[*]`
* Field not present in the structure `someObject.fieldThatDoesNotExist`
* Array/object operations on simpleTypes
* A test for recursive shapes

Reply:
Added more linter tests